File: server/app/schemas/progress.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date, time, datetime
from app.schemas.subject import SubjectResponse, ChapterResponse

class StudyLogBase(BaseModel):
    subject_id: int
    chapter_id: Optional[int] = None
    log_date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    duration_minutes: int = Field(..., gt=0)
    topics_covered: Optional[str] = None
    notes: Optional[str] = None
    mood: Optional[int] = Field(None, ge=1, le=5)
    energy_level: Optional[int] = Field(None, ge=1, le=5)

class StudyLogCreate(StudyLogBase):
    pass

class StudyLogUpdate(BaseModel):
    subject_id: Optional[int] = None
    chapter_id: Optional[int] = None
    log_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    duration_minutes: Optional[int] = Field(None, gt=0)
    topics_covered: Optional[str] = None
    notes: Optional[str] = None
    mood: Optional[int] = Field(None, ge=1, le=5)
    energy_level: Optional[int] = Field(None, ge=1, le=5)

class StudyLogResponse(StudyLogBase):
    id: int
    user_id: int
    created_at: datetime
    
    subject_name: Optional[str] = None
    subject_color: Optional[str] = None
    chapter_title: Optional[str] = None

    class Config:
        from_attributes = True

class DailyProgressSummary(BaseModel):
    log_date: date
    total_minutes: int
    logs_count: int

class SubjectProgressResponse(BaseModel):
    subject_id: int
    subject_name: str
    color: str
    total_estimated_hours: float
    total_completed_hours: float
    progress_percentage: int

File: client/src/types/progress.ts
export interface StudyLog {
  id: number;
  subject_id: number;
  chapter_id?: number;
  log_date: string;
  start_time?: string;
  end_time?: string;
  duration_minutes: number;
  topics_covered?: string;
  notes?: string;
  mood?: number;
  energy_level?: number;
  created_at: string;
  
  subject_name?: string;
  subject_color?: string;
  chapter_title?: string;
}

export interface StudyLogCreate {
  subject_id: number;
  chapter_id?: number | null;
  log_date: string;
  duration_minutes: number;
  topics_covered?: string;
  notes?: string;
  mood?: number;
  energy_level?: number;
}

export interface DailyProgressSummary {
  log_date: string;
  total_minutes: number;
  logs_count: number;
}

export interface SubjectProgress {
  subject_id: number;
  subject_name: string;
  color: string;
  total_estimated_hours: number;
  total_completed_hours: number;
  progress_percentage: number;
}

export interface ProgressOverview {
  total_study_hours: number;
  global_progress_percentage: number;
  current_streak: number;
  subject_progress: SubjectProgress[];
}
